File: C/18_Stack/BOJ_10799_쇠막대기.c
#include <stdio.h>
#include <string.h>
int stack[100000];
int idx = -1;
void push(int value) {
	stack[++idx] = value;
}
int pop() {
	if (idx < 0) return -1;
	else return stack[idx--];
}
int size() {
	return idx + 1;
}
main() {
	int sum = 0;
	char str[100001];
	scanf("%s", str);
	for (int i = 0; i < strlen(str); i++) {
		if (str[i] == '(') {
			if (str[i + 1] == ')') {
				sum += size();
				i++;
			}
			else push(1);
		}
		else {
			sum++;
			pop();
		}
	}
	printf("%d", sum);
}

File: JAVA/18_Stack/BOJ_10799_쇠막대기.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Stack;

public class Main {
    public static void main(String[] args) throws IOException {
        Stack<Integer> stack = new Stack<Integer>();
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String str = br.readLine();
        int sum = 0;
        for (int i = 0; i < str.length(); i++) {
            if (str.charAt(i) == '(') {
                stack.push(1);
            } else {
                if (!stack.empty() && str.charAt(i - 1) == '(') {
                    stack.pop();
                    sum += stack.size();
                } else {
                    stack.pop();
                    sum++;
                }
            }
        }
        System.out.println(sum);
    }
}